core/src/main/scala/kafka/coordinator/transaction/TransactionMetadata.scala

```diff
@@ -0,0 +1,167 @@
+ * Transaction has not existed yet
+ *
+ * transition: received AddPartitionsToTxnRequest => Ongoing
+ *             received AddOffsetsToTxnRequest => Ongoing
+ */
+private[coordinator] case object Empty extends TransactionState { val byte: Byte = 0 }
+
+/**
+ * Transaction has started and ongoing
+ *
+ * transition: received EndTxnRequest with commit => PrepareCommit
+ *             received EndTxnRequest with abort => PrepareAbort
+ *             received AddPartitionsToTxnRequest => Ongoing
+ *             received AddOffsetsToTxnRequest => Ongoing
+ */
+private[coordinator] case object Ongoing extends TransactionState { val byte: Byte = 1 }
+
+/**
+ * Group is preparing to commit
+ *
+ * transition: received acks from all partitions => CompleteCommit
+ */
+private[coordinator] case object PrepareCommit extends TransactionState { val byte: Byte = 2}
+
+/**
+ * Group is preparing to abort
+ *
+ * transition: received acks from all partitions => CompleteAbort
+ */
+private[coordinator] case object PrepareAbort extends TransactionState { val byte: Byte = 3 }
+
+/**
+ * Group has completed commit
+ *
+ * Will soon be removed from the ongoing transaction cache
+ */
+private[coordinator] case object CompleteCommit extends TransactionState { val byte: Byte = 4 }
+
+/**
+ * Group has completed abort
+ *
+ * Will soon be removed from the ongoing transaction cache
+ */
+private[coordinator] case object CompleteAbort extends TransactionState { val byte: Byte = 5 }
+
+private[coordinator] object TransactionMetadata {
+  def apply(pid: Long, epoch: Short, txnTimeoutMs: Int, timestamp: Long) = new TransactionMetadata(pid, epoch, txnTimeoutMs, Empty, collection.mutable.Set.empty[TopicPartition], timestamp)
+
+  def apply(pid: Long, epoch: Short, txnTimeoutMs: Int, state: TransactionState, timestamp: Long) = new TransactionMetadata(pid, epoch, txnTimeoutMs, state, collection.mutable.Set.empty[TopicPartition], timestamp)
+
+  def byteToState(byte: Byte): TransactionState = {
+    byte match {
+      case 0 => Empty
+      case 1 => Ongoing
+      case 2 => PrepareCommit
+      case 3 => PrepareAbort
+      case 4 => CompleteCommit
+      case 5 => CompleteAbort
+      case unknown => throw new IllegalStateException("Unknown transaction state byte " + unknown + " from the transaction status message")
+    }
+  }
+
+  def isValidTransition(oldState: TransactionState, newState: TransactionState): Boolean = TransactionMetadata.validPreviousStates(newState).contains(oldState)
+
+  private val validPreviousStates: Map[TransactionState, Set[TransactionState]] =
+    Map(Empty -> Set(),
+      Ongoing -> Set(Ongoing, Empty, CompleteCommit, CompleteAbort),
+      PrepareCommit -> Set(Ongoing),
+      PrepareAbort -> Set(Ongoing),
+      CompleteCommit -> Set(PrepareCommit),
+      CompleteAbort -> Set(PrepareAbort))
+}
+
+@nonthreadsafe
+private[coordinator] class TransactionMetadata(val pid: Long,
+                                               var epoch: Short,
```

Comment:
Actually, i think this should be `producerEpoch`? AFAIK `coordinatorEpoch` is `partition.getLeaderEpoch`